cloudinit/sources/DataSourceCloudStack.py

```diff
@@ -208,38 +211,43 @@ def _get_data(self):
             LOG.debug("Using seeded cloudstack data from: %s", self.seed_dir)
             return True
         try:
+            if self.perform_dhcp_setup:
+                LOG.debug("Setting up network to reach IMDS")
+                candidate_nics = net.find_candidate_nics()
+                LOG.debug("Looking for the primary NIC in: %s", candidate_nics)
+
+                for candidate_nic in candidate_nics:
+                    network_context = EphemeralIPNetwork(
+                        self.distro,
+                        candidate_nic,
+                    )
+                    try:
+                        with network_context:
```

Comment:
Suggestion using net.find_fallback_nic() as mentioned above

```suggestion
                try:
                    primary_nic = net.find_fallback_nic()
                    with EphemeralIPNetwork(self.distro, primary_nic):
                        vr_addr = get_vr_address(self.distro)
                        # If vr_addr is a dict, we have the DHCP lease
                        self.vr_addr = (
                            vr_addr.get("dhcp-server-identifier")
                            if isinstance(vr_addr, dict)
                            else vr_addr
                         )
                        self.metadata_address = f"http://{self.vr_addr}/"

                        return self._crawl_metadata()
                except NoDHCPLeaseError:
                    LOG.debug(
                        "Unable to obtain a DHCP lease for %s",
                        primary_nic,
                    )
                except Exception:
                    LOG.debug(
                        "Failed to crawl IMDS with %s", primary_nic
                    )
```